README.md
# RandPassGenerator
RandPassGenerator 1.3.3

The RandPassGenerator Java application is a simple command-line utility for generating random passwords, passphrases, and raw keys. It is designed very conservatively to ensure that the random values it provides offer full cryptographic strength requested by the user. 


### Build Information

To build the RandPassGenerator jar file, use the Oracle Java SDK; any recent version should be fine.

Go to the directory containing build.xml, and run ant as shown below.

	ant jar

This will create build/jar/PassGenerator.jar.  Copy the jar file to somewhere convenient where you want to generate passwords or passphrases.

	cp build/jar/PassGenerator.jar $HOME


### Usage Information


To use RandPassGenerator, you'll need the Oracle Java Runtime Environment; any recent version should be sufficient, but at a minimum version 9 is recommended.

The RandPassGenerator can run from a terminal or console. The command-line syntax is simple:

	java -jar PassGenerator.jar  [options]

### Options

-v	  {Print verbose messages during operation, in addition to logging}

-str S    {Use generation strength of S bits (default: 160)}

-pw N	  {Generate N random password of the specified strength}

-pp N	  {Generate N random passphrases of the specified strength}

-k N	  {Generate N random keys of the specified strength}

-enc	   {Encrypt generated random key using a random password that is at least a 16 characters (256-bit AES) and write to file named the Key ID (KEY_ID.enc). A prompt for a random password to us will appear. Users should generate a random password to use for encryption prior to generating keys. ("java -jar PassGenerator.jar -pw 1 -str 96" will generate a 16 character password).}

-decrypt   {Decrypt encrypted key file using a random password that is at least a 16 characters and save as text file (KEY_ID_decrypted.txt). A prompt for the name of the encrypted file to decrypt will appear, then a prompt for the random password to use will appear.}

Unusual options:
  
-pplen M  {When generating passphrases, longest word should be M letters long (minimum value of M is 3)}

-ppurl U  {Use the URL U to load words for passphrase (default: use internal list).  Words must be at least 3 letters long.}

-pwcs P   {Use character pattern P for characters to use in passwords (lowercase, uppercase, number, special character, or combination)}

-pwcustom F {Use the specified file F as the source of a custom character set; F must be readable}

-log F    {Log all operations to the log file F (default: ./randpass.log)}

-out F    {Write output to file F (default: writes to stdout)}

-c N 	    {Format output passwords and keys in chunks of N characters}

-sep S    {For chunk formatting, use S as the separator (default: -)}

-rcc N    {For passphrases - impose random camel-case; randomly uppercase the first N letters (default: 0)}

At least one of the options -pw, -pp, or -k must be supplied. The keys, passwords, or passphrases produced by RandPassGenerator will be written to the standard output (stdout), so they can easily be redirected to a file. The -out option can also be used to write the output to a file. All messages are written to the standard error (stderr).

Detailed log messages are appended to the specified log file - if the log file cannot be opened, then the tool will not run. 

Note that the -pwcs option is a little strange. Each character in the value represents a full set of characters. Any lowercase letter
means "add a character set of all lowercase letters", any uppercase letter means "add a set of all uppercase letter", any digit means 
"add a set of all digits", and anything else means "add a set of all punctuation marks".  Normally, you should not use the -pwcs option, you should let RandPassGenerator use its default character set.

If you want a fully custom character set, use the -pwcustom option.  For this option, you provide a file.  Each printable character in the file is taken as a character for a custom password character set.  Non-printable characters like TAB or NEWLINE are ignored.  Note that the set is de-duped, so even if the letter 'A' appears six times, it acts as if it appeared once.  The -pwcustom and -pwcs options may not be used together, at most one of them may appear for a given invocation of RandPassGenerator.

The random camel case option (-rcc N) applies only when generating passphrases using the -pp option.  Using -rcc N will apply uppercase at 50% chance to the first N letters of each passphrase word.  By default the value for this option is 0, which means that no uppercasing will be applied.  For a value of 1, only the first letter of each word might be transformed to uppercase, for 2, only first and second letter, etc.  Note that camel case can add entropy to the passphrase, but that the entropy strength does NOT take camel case into account because it varies too much.


### Examples

Example 1: generate 5 random passwords using the default mixed character set, at default strength of 160, saved into file GoodPasswords.dat
 
	java -jar PassGenerator.jar -pw 5 >GoodPasswords.dat

Example 2: generate 20 random passphrases using the default dictionary, at strength of 256, with verbose messages, using words up to 9 letters long, and output saved into the file passphrases.txt

	java -jar PassGenerator.jar -v -pp 20 -str 256 -pplen 9 >passphrases.txt

Example 3: generate 200 random keys at strength of 192, with logging to keygen.log, and output to mykeys.out.

	java -jar PassGenerator.jar -k 200 -str 192 -log keygen.log -out mykeys.out

Example 4a: generate 100 passwords at strength 160, using a character set of lowercase letters and digits, with output redirected to hi-quality-stuff.txt
    
	java -jar PassGenerator.jar -pw 100 -pwcs "a0"  >hi-quality-stuff.txt

Example 4b: generate 100 passwords at strength 96, using a custom character set, and verbose output messages

	java -jar PassGenerator.jar -pw 100 -str 96 -pwcustom MyPwdChars.txt -v

Example 5: generate 10 passwords at strength 128, formatted into chunks of five characters each, separated by /.

	java -jar PassGenerator.jar -pw 10 -str 128 -c 5 -sep /

Example 6: generate 1 random key at strength 256, and encrypt to file using random password.

	java -jar PassGenerator.jar -k 1 -str 256 -enc

Example 7: Decrypt encrypted key file.  

	java -jar PassGenerator.jar -decrypt
	
Example 8: generate 6 passphrases at strength ~100, but using base strength of 94 plus random upcase of first letter

	java -jar PassGenerator.jar -pp 6 -pplen 7 -str 94 -rcc 1


### Design Information
The foundation of RandPassGenerator is an implementation of the NIST SP800-90 Hash DRBG.  It uses entropy, carefully gathered from system sources, to generate quality random output.  The internal strength of the DRBG is 192 bits, according to NIST SP800-57, using the SHA-384 algorithm. In accordance with SP800-90, the DRBG is seeded with at least 888 bits of high quality entropy from entropy sources prior to any operation.
 
This implementation uses the seed mechanism of the Java SecureRandom class for gathering entropy. This implementation performs self-tests at every execution, so that users can be confident that no library problems have affected operation. Two kinds of self-tests are performed:

1. Known-answer tests from the NIST Hash_DRBG verification suite test file.
2. Simple statistical tests on DRBG output.

If the tests don't pass, the tool reports failure and refuses to run. 

The strength mechanism implemented here is quite simple. For passwords, the size of the character set used defines the 
bits-per-character, and password length is then computed to meet or exceed the requested strength (typically, this is somewhere around 5-6 bits per character). Similarly, for passphrases the size of the usable dictionary defines the bits-per-word, and passphrase length is then computed to meet or exceed the requested strength (for the default dictionary and settings, roughly 16 bits-per-word). Duplicates are eliminated and the entropy is computed based on the number of unique characters or words. 

The RandPassGenerator tool performs extensive logging. By default, log entries are appended to the local file "randpass.log". No actual key data, random data, or seed data is written to the log file.

## License

See [LICENSE](./LICENSE.md).

## Contributing

See [CONTRIBUTING](./CONTRIBUTING.md).

## Disclaimer

See [DISCLAIMER](./DISCLAIMER.md).




RandPassGenerator/CHANGELOG.txt

CHANGES FROM 1.0.1  TO  1.1.0


* Added static method to AbstractDRBG to correctly convert from an unsigned
  Java BigInteger to a byte array. 
Then applied this method everywhere.
  Also changed construction of all BigIntegers to ensure they were created
  as unsigned when being constructed from a byte array.


* Added code to CharacterSet class to count entropy based on unique characters
  in the set.  This allows for a passwords with correct strength even if the
  CharacterSet has repeated characters.


* Added code to CharacterSet to reset the internal random number buffer when
  character set is rebuilt.  This was unnecessary, since the buffer 16-bit 
  integer contents are independent of character set, but the reviewers specified
  it should be done.


* Added code to the WordSet class to remove duplicate words while loading the
  words from the source URL or internal list.  This ensures passphrases of
  correct strength even if the input had redundant words.


* Disabled re-use of self-test entropy with the HashDRBG (code in the RandManager
  class).  This is based on reviewers' interpretation of SP800-90A sect 11.3.2
  prohibiting re-use of self-test entropy in any form.  My own interpretation
  of 11.3.2 is different, but in deference to the reviewers RandManager now
  discards perfectly good system entropy rather than add it to the Nonce.


* Added code to HashDRBG class to check for minimum entropy input size, for better
  conformance to SP800-90, in instantiate() and reseed().


* Added logging to HashDRBG class to record the entropy source being used, after
  the source passes self-test.


* Added code to the toString method of the JavaSecRandEntropySource to include
  some info about the underlying Java SecureRandom objects, to aid with logging.

* Added new option to write key/passwd/passphrase output to a named file.  This
  was also unnecessary, since stdout redirection can do the same thing, but
  it is a pretty common convention for command line utilities.


* Added README.txt file with examples.




CHANGES FROM  1.1.0  TO  1.1.2


* Added new class FixedValuesEntropySource to support known answer testing.
  This class is used to supply specific values of entropy input to drive tests.


* Added new class HashDRBGKnownAnswerTest to support known answer testing.


* Added code to the HashDRBG class to support known answer testing, and fixed
  several small buffer size bugs that prevented getting the known answers.


* Added two specific NIST known answer tests, one for SHA-256 and one for
  SHA-384.


* Added code to the RandManager startup to ensure that the HashDRBG known
  answer tests were invoked and passed before using the HashDRBG.


* Fixed interpretation of the strength values in the HashDRBG class and
  elsewhere to conform more closely to SP800-90/57/107  [NOTE: while nominal
  'strength' of SHA-384 is 192 bits, that is based on the collision resistance,
  but for the HashDRBG key generation use case, the 'strength' should be
  taken from the pre-image resistance, with is nominally 384 bits.  So, we
  allow the HashDRBG built on SHA-384 to generate keys with strength up to
  384 bits.]


* Added more examples to README.txt file.



CHANGES FROM  1.1.2  TO  1.1.3


* Added formatting options to permit more readable output formatting of
 hex keys and passwords.



CHANGES FROM  1.1.3  to  1.1.4


* Enforced de-duped character set for all operations, including
  generation, in gen/CharacterSet.java, i.a.w. recommendations from
  the reviewers.
  

* Minor changes to README.txt file.



CHANGES FROM  1.1.4  to  1.1.5
* Added code to HexKeyGen class to enable key generation error messages to print to command line
* Added code to RandPassGenerator class to enable error messages to print to command line without verbose '-v' command
* Added code to CharacterSet class to enable password generation error message to print to command line for 0 or negative strength values
* Added code to WordSet class to enable passphrase generation error message to print to command line for 0 or negative strength values
* Minor changes to README.txt file.



CHANGES FROM  1.1.5  to  1.2
*Updated code in RandPassGenerator and RandManager classes to disable saving of entropy to a file at shutdown. Instead of saving entropy to a file during shutdown for later use as the nonce entropy source or using the system time when an entropy file doesn't exist, the same entropy source as the primary entropy input for the DRBG seed is used for the nonce. SP800-90A 8.6.7 states nonce should be a random value with at least (security_strength/2) bits of entropy and the value could be acquired from the same source and at the same time as the entropy input.
*Added Key ID Generator to RandPassGenerator class for key accounting. It generates a KeyID for generated keys using first 64-bits of a hash of the key and current date (YYYYMMDD_hhmmssH1H2H3H4H5H6H7H8H9H10H11H12H13H14H15H16). 
*Added key generation transaction log to RandPassGenerator class.
*Added FileEncryption and FileDecryption utilities to allow for output files containing keys to be encrypted and unencrypted using AES and a random password. 
*Updated README.txt file.

CHANGES FROM  1.2  to  1.3
*Added KeyWrapper and KeyUnwrapper utilities to allow for output files containing keys to be encrypted and unencrypted using AES Key Wrap (NIST SP 800-38F/RFC 3394) and a random password. DPKDF2 is used in these classes as specified in NIST SP 800-132 to derive a 256-bit key encryption key (KEK) from a user inputted random password.  

RandPassGenerator/src/gov/nsa/ia/pass/RandManager.java
package gov.nsa.ia.pass;

import gov.nsa.ia.drbg.*;
import gov.nsa.ia.util.*;


import java.util.logging.Level;
import java.util.logging.Logger;

/**
 * This class manages creation and maintenance of the
 * DRBGs and EntropySource objects necessary for
 * generating cryptographic keys, passwords, and passphrases.
 * 
 * In addition to managing the entropy sources, this
 * class also has the ability to, at startup, read
 * extra entropy from a file, and at shutdown to
 * save entropy to a file, if configured by its
 * caller to do so.  
 * 
 * This class manages three types of entropy sources.
 * <ol>
 * <li>The primary entropy source is used as the source
 * for a HashDRBG.  This must be an really solid source
 * that generates very high entropy-per-byte, can be
 * called repeatedly, and blocks until it has high
 * quality entropy to deliver.  Normally this would be
 * a UnixDevRandomEntropySource on Linux.</li>
 * <li>A supplemental entropy source is used for 
 * additional input for one generation call
 * per regrab interval (normally 1 minute).  It
 * must be callable repeatedly, but does not need
 * to deliver high entropy-per-byte.  This is
 * optional but recommended</li>
 * <li>A startup entropy source is used for the
 * nonce value for HashDRBG instantiation, and is
 * also used to save some high-quality entropy from
 * the HashDRBG at shutdown.  This should be
 * a FileEntropySource; it must be callable once
 * at startup (getEntropy method) and once at 
 * shutdown (saveEntropy method).
 * </ol>
 * 
 * The usual calling sequence for this class would be:
 * create object, create sources, set sources, set logger,
 * call initialize method, call performSelfTest method,
 * call generateKey method many times, call shutdown method.
 * Also, it is often a good idea to set up a ShutdownHook
 * to call the shutdown method, by using Runtime.addShutdownHook
 * and the utility method getKeyGenerationShutdownHook().
 * 
 * @author nziring
 * Updated 09252018 by amsagos
 */
public class RandManager implements SelfTestable {
    /**
     * Number of blocks of good entropy to save to a file
     * at shutdown, if possible.  Block size is 8 bytes.
     */
    public static final int SAVE_BLOCKS = 16;
	
    /**
     * Basic strength to request for the DRBG, not a good idea
     * to override this, so this class provides no
     * way to do so.  This uses the maximum recommended
     * hash size of the DRBG under the NSAS: SHA-384.
     */
    public static final int DRBG_STRENGTH = 192;
	
    /**
     * Amount of entropy we can reasonably expect from 
     * the supplemental source, per call, in bits.
     */
    public static final int SUPPLEMENT_ENTROPY = 128;
	
    /**
     * Max amount of entropy we can unreasonably expect from the
     * startup source, per call, in bits.  (Remember, we believe
     * the startup source is extremely high-quality flat random
     * bits, so 512 bits of entropy from it should require only
     * 64 bytes of data.)
     */
    public static final int STARTUP_SOURCE_ENTROPY = 512;

    /**
     * Set to true if primary source self-test entropy may
     * be added to the nonce given to the instantiate function
     * of the DRBG.  
     * 
     * POST-REVIEW NOTE: Default value is false because SP800-90A
     * 11.3.2 can be construed as prohibiting re-use of self-test
     * entropy in any form.  Which is a darn shame since it is
     * perfectly good entropy.  (My own opinion is that the words
     * in 11.3.2 were intended to prohibit re-use of self-test
     * entropy as the primary entropy input, rather than as part
     * of the nonce.  Oh well.)
     */
    public static final boolean USE_EXTRA_SELFTEST_ENTROPY = false;


    private EntropySource primarySource;
    private EntropySource suppSource;
    private EntropySource startupSource;
    private HashDRBG rbg;
    private boolean okay;
    private String name;
    private Logger log;
    //private long lastSupplementTime;
	
    /**
     * Constructor - create a KeyGenerationManager object.
     * The object will not yet be usable.  At the least, 
     * you must call setPrimarySource() to supply a usable
     * primary EntropySource and initialize() to fire up
     * the DRBG and perform successful self-test.
     * 
     * @param nam Base name for this manager, must be non-null
     * @param l Logger to use; if null, then a new Logger will be created
     */
    public RandManager(String nam, Logger l) {
	primarySource = null;
	suppSource = null;
	startupSource = null;
	okay = false;
	if (nam == null) throw new IllegalArgumentException("Non-null Name must be supplied.");
	name = nam;
	setLogger(l);
	log.fine("Initialized RandManager object, name=" + name);
    }
	
	
    /**
     * Return our Logger, in case caller needs to set something
     * on it or log some messages.  This will always return
     * a non-null, usable Logger.
     */
    public  Logger getLogger() {
	return log;
    }
	
    /** 
     * Set a new Logger for us.  This must be done BEFORE
     * calling initialize(), or it will be useless.  If
     * this is called with null, it will cause a new default
     * Logger to be created, but thats not usually a good idea.
     * 
     * @param l Logger to use for logging our operations
     */
    public  void setLogger(Logger l) {
	if (l != null) {
	    log = l;
	} else {
	    log = Logger.getLogger(name + "-log");
	    log.setLevel(Level.FINE);
	}
	if (rbg != null) {
	    rbg.setLogger(log);
	}
    }

    /**
     * Perform a shutdown of this RandManager.
     * This may only be called once, after which this
     * object is not usable.  This method has three jobs:
     * (1) save entropy to the startupSource if set, and
     * (2) call uninstantiate on our DRBG, and
     * (3) call dispose on all our entropy sources.
     */
   public  void shutdown() {
	// step 1
	//if (okay && (rbg != null) && (startupSource != null)) {
	  //  try {
		//if (startupSource.saveEntropy(rbg, SAVE_BLOCKS)) {
		//    log.info("In RandManager " + name + " shutdown, saved entropy to " +
		//	     startupSource.getDestination());
		//} else {
		//    log.warning("In RandManager " + name + ", attempt to save entropy failed.");
		//}
	   // } catch(IOException ie) {
		//log.warning("Saving entropy to file failed: " + ie);
	  //  }
	//} else {
	 //   log.warning("In shutdown, unable to save entropy to file.");
	//}
		
	// step 2
	if (rbg != null && rbg.isOkay()) {
	    rbg.uninstantiate();
	    rbg = null;
	}
	okay = false;
		
	// step 3
	if (primarySource != null) {
	    primarySource.dispose();
	    primarySource = null;
	}
	if (suppSource != null) {
	    suppSource.dispose();
	    suppSource = null;
	}		
	if (startupSource != null) {
	    startupSource.dispose();
	    startupSource = null;
	}
		
	log.info("Shut down RandManager " + name);
	return;
    }
	
    /**
     * Set the primary entropy source.  This source must
     * be callable repeatedly, and must block until it has
     * high-quality entropy to deliver.  This method
     * must be called before calling initialize(). 
     * If a primary source was
     * already set, that source will have its dispose()
     * method called prior to being replaced.
     * 
     * @param src A good EntropySource, such as a UnixDevRandomEntropySource or JavaSecRandEntropySource.
     */
    public  void setPrimarySource(EntropySource src) {
	if (src == null) {
	    throw new IllegalArgumentException("Primary entropy source must be non-null.");
	}
	if (src == primarySource) return; // No-op
		
	if (primarySource != null) {
	    primarySource.dispose();
	}
	primarySource = src;
	log.info("Set RandManager " + name + " primary entropy source to " + src);
	return;
    }

    /**
     * Set the supplementary entropy source.  This source must
     * be callable repeatedly.  This method
     * should be called before initialize(), but
     * may be called again later to set a different
     * supplemental source if desired.  If a supplemental source
     * is already set, that source will have its dispose()
     * method called prior to being replaced.  It is okay to
     * not set a supplemental source.
     * 
     * @param src An EntropySource, such as a FileEntropySource.
     */
    public  void setSupplementalSource(EntropySource src) {
	if (src == suppSource) return;	
	if (suppSource != null) {
	    suppSource.dispose();
	}
	suppSource = src;
	if (src != null) {
	    log.info("Set RandManager " + name + " supplemental entropy source to " + src);
	} else {
	    log.info("Set RandManager " + name + " supplemental entropy source to none.");
	}
	return;
    }
	
    /**
     * Set the startup entropy source.  This source must
     * be a FileEntropySource.  This method
     * should be called before initialize().
     * 
     * @param src A FileEntropySource that can support saveEntropy()
     */
    //public  void setStartupSource(FileEntropySource src) {
    public  void setStartupSource(EntropySource src) {
	if (src == startupSource) return;	
	if (startupSource != null) {
	    startupSource.dispose();
	}
	startupSource = src;
	if (src != null) {
	    log.info("Set RandManager " + name + " start/save entropy source to " + src);
	} else {
	    log.info("Set RandManager " + name + " start/save entropy source to none.");
	}
	return;
    }
	
    /**
     * Return true if this RandManager is in a usable state.
     */
    public  boolean isOkay() {
	return okay && (rbg != null) && (rbg.isOkay());
    }
	
    /**
     * Initialize this RandManager object.  This method should
     * be called exactly once, before this RandManager can be
     * used.  If initialization succeeds, this method returns true
     * and the method isOkay() will return true also.  
     * 
     * Internally, this method uses the primarySource and the startupSource.
     * It builds the SP800-90 nonce from the startupSource and self-test
     * entropy from the primarySource.
     * 
     * @return true if initialization succeeds, false otherwise
     */
    public  boolean initialize() {
	byte[] nonce = null;
	byte[] testEntropy = null;
	int status;
		
	if (isOkay()) return true;
		
	// check pre-conditions
	if (primarySource == null) {
	    log.warning("Could not initialize RandManager " + name + " because no primary source has been set.");
	    return false;
	}
		
	// attempt to grab DRBG nonce from startupSource if possible, or system time otherwise
	// 1.2 UPDATE: See note in RandPassGenerator line 147.
	if (startupSource != null) {
	    nonce = startupSource.getEntropy(STARTUP_SOURCE_ENTROPY, 8, 256);
	    if (nonce != null && nonce.length > 0) {
		//log.fine("Got startup entropy nonce from " + startupSource.getDestination() + ", " + nonce.length + " bytes.");
	    log.fine("Got startup entropy nonce from " + "primarySource" + ", " + nonce.length + " bytes.");
	    }
	}
	if (nonce == null) {
	    nonce = (System.currentTimeMillis() + "x").getBytes();
	 log.fine("Got startup entropy nonce from system time."+ ", " + nonce.length + " bytes.");
	}
		
	// append self-test entropy from the primary source to the nonce, 
	// if allowed.
	// POST-REVIEW NOTE: this is prohibited by default, see above.
	//
	if (USE_EXTRA_SELFTEST_ENTROPY) {
	    testEntropy = primarySource.getSelfTestEntropy();
	    if (testEntropy != null && testEntropy.length > 0) {
		log.fine("Got additional startup nonce entropy from primary source self-test, "
			 + testEntropy.length + " bytes.");
		byte [] newnonce = new byte[nonce.length + testEntropy.length];
		System.arraycopy(nonce, 0, newnonce, 0, nonce.length);
		System.arraycopy(testEntropy, 0, newnonce, 0, testEntropy.length);
		nonce = newnonce;
	    }
	}
		
	// create the DRBG and instantiate it, this will cause the
	// DRBG to invoke the self-test of the primary source.
	rbg = new HashDRBG(name + "-drbg", primarySource);
	rbg.setLogger(log);
	status = rbg.instantiate(DRBG_STRENGTH, false, name, nonce);
	if (status == DRBGConstants.STATUS_SUCCESS) {
	    log.info("RandManager " + name + " successfully instantiated DRBG: " + rbg);
	} else {
	    log.severe("Instantiation of RandManager DRBG failed.  Cannot generate keys in " + name);
	    return false;
	}
		
	// all done!
	okay = true;
	return okay;
    }
	
	
    /**
     * Return the current DRBG.  Returns null if
     * this RandManager is not okay.
     *
     * Note that the caller should NOT uninstantiate the returned
     * DRBG.  The correct thing to do is to call shutdown on this
     * RandManager when all use of the DRBG is done.
     *
     * @return a HashDRBG, already instantiated, or null
     */
    public AbstractDRBG getDRBG() {
	if (!isOkay()) return null;

	return rbg;
    }

    
    /**
     * Generate a key of a specified size and return it.
     * If we cannot generate a key, return null.  This
     * method uses the generate() method of the DRBG.
     * If the time since the most recent use of the
     * supplemental source was more than the supplement
     * interval, then we grab some additional input from
     * the supplemental source and use it with the call
     * to the DRBG generate method.
     * 
     * @param numBytes size of the desired random key in bytes, more than 0
     */
    public  byte[] generateKey(int numBytes) {
	byte [] ret = null;
	byte [] addlInput = null;
	int status;
		
	// check that is generator is okay
	if (!isOkay()) {
	    log.warning("Attempt to call RandManager " + name + " generation when not okay!");
	    return null;
	}
		
	// validate that the value of numBytes is positive
	if (numBytes <= 0) {
	    log.warning("Attempt to call RandManager " + name +
			" generation with size 0 or negative.  Null returned.");
	    return null;
	}
		
	// validate that the value of numBytes is not larger than the 
	// instantiated DBRG strength
	if (numBytes > (rbg.getStrength() / 8)) {
	    log.warning("Attempt to call RandManager " + name + 
			" generation for+size " + numBytes + " which is greater than " +
			" available DRBG strength.  Null returned.");
	    return null;
	}
		
	// perform generation
	ret = new byte[numBytes];
	log.fine("RandManager " + name + " about to call DRBG generate for " + numBytes + " bytes.");
	status = rbg.generate(numBytes, 0, false, addlInput, ret);
	if (status == DRBGConstants.STATUS_SUCCESS) {
	    log.fine("RandManager " + name + " generated key of bytes " + ret.length);
	} else {
	    log.severe("RandManager " + name + " key generation failed!");
	    ret = null;
	}
		
	// all done
	return ret;
    }
	
	
    private static final int SELF_TEST_SIZES[] = { 16, 32 };

    /**
     * Perform a basic Known Answer test of our underlying DRBG.
     * This calls the static method on the HashDRBG class, and simply
     * returns the result, true or false.  This should be performed, and
     * should pass, before you call initialize().  
     *
     * @return true if HashDRBG known answer self-test passed
     */
    public boolean performDRBGKATest() {
	boolean result = HashDRBG.performKnownAnswerTests(getLogger());
	return result;
    }
	


    /**
     * Perform a self-test on a properly initialized RandManager object.
     * This tests generation of keys of sizes SELF_TEST_SIZES. All the test
     * generations must succeed in order for self-test to pass. On pass, this
     * method returns true.
     */
    public boolean performSelfTest() {
	if (!isOkay()) {
	    log.info("Self-test failing due to generator state not okay.");
	    return false;
	}

	log.fine("About to perform self-test on HashDRBG class.");
	HashDRBG trbg;
	trbg = new HashDRBG("self-test", new LousyEntropySource());
	trbg.setLogger(log);
	if (!trbg.performSelfTest()) {
	    log.warning("HashDRBG class failed self-test.  Failing RandManager self-test.");
	    return false;
	}

	log.info("Self-test of HashDRBG class succeeded, continuing to self-test of RandManager "
		 + name);
	byte[] keyval;
	int siz, i;
	for (i = 0; i < SELF_TEST_SIZES.length; i++) {
	    siz = SELF_TEST_SIZES[i];
	    log.fine("Self-test generating a raw key at size " + siz);
	    keyval = generateKey(siz);
	    if ((8*siz) <= rbg.getStrength()) {
		if (keyval == null) {
		    log.warning("Self-test failing at key size " + siz
				+ ", generateKey returned null.");
		    return false;
		}
		if (keyval.length != siz) {
		    log.warning("Self-test failing at key size " + siz
				+ ", generateKey returned wrong size.");
		    return false;
		}
	    } else {
		if (keyval != null) {
		    log.warning("Self-test failing, generateKey returned a value for size " + siz +
				" when the size was actually greater than available DRBG strength.");
		    return false;
		}
	    }
	    log.fine("Self-test for size " + siz + " succeeded.");
	}

	return true;
    }


    // TESTING
    
    protected static final int TEST_KEY_COUNT = 10;
    protected static final int TEST_KEY_SIZE = 32;
	
    /**
     * @param args
     */
    public static void main(String[] args) {
	String suppFile = null;
	String saveFile;
	RandManager kgm;
	EntropySource primarysrc;
	FileEntropySource startupsrc;
	FileEntropySource suppsrc = null;
		
	if (args.length == 0) {
	    System.err.println("Usage: java gov.nsa.ia.RandManager startup-file [suppl-file]");
	    System.exit(0);
	}
		
	saveFile = args[0];
	if (args.length > 1) suppFile = args[1];
		
	try {
	    kgm = new RandManager("UnitTestKGM", Logger.getLogger("RandManTest"));
	    kgm.getLogger().setLevel(Level.FINE);
	    primarysrc = new JavaSecRandEntropySource();
	    // primarysrc = new UnixDevRandomEntropySource();
	    startupsrc = new FileEntropySource(saveFile, 8, 2);
	    if (suppFile != null) suppsrc = new FileEntropySource(suppFile, 8, 2);
	    System.err.println("Created sources: p=" + primarysrc + ", s=" + startupsrc + ", s=" + suppsrc);
	    kgm.setPrimarySource(primarysrc);
	    kgm.setStartupSource(startupsrc);
	    kgm.setSupplementalSource(suppsrc);
	    System.err.println("About to initialize, might take time..");
	    if (kgm.initialize()) {
		System.err.println("KGM initialization succeeeded, proceeded to self-test, this might take a while if the DRBG blocks.");
		if (kgm.performSelfTest()) {
		    System.err.println("KGM self-test passed!");
		} else {
		    System.err.println("KGM self-test failed.  Bummer!");
		}
				
		// now try to get some test keys using the DRBG
		int ki;
		byte [] testkey;
		kgm.getLogger().setLevel(Level.INFO);
		for(ki = 0; ki < TEST_KEY_COUNT; ki++) {
		    testkey = kgm.generateKey(TEST_KEY_SIZE);
		    if (testkey == null || testkey.length != TEST_KEY_SIZE) {
			System.err.println("Key generate failed on bulk test iteration " + ki);
			System.exit(1);
		    }
		    // possibly print some of the keys here
		    System.err.println("Bulk test iteration no. " + ki + " key is " + EntropyUtil.bufferToString(testkey));
		    System.err.println("\tentropy is " + EntropyUtil.computeByteEntropy(testkey, TEST_KEY_SIZE));

		}
				
		System.err.println("Performing KGM shutdown.");
		kgm.shutdown();
	    } else {
		System.err.println("KGM initialization failed!");
	    }	
	} catch (Exception e) {
	    System.err.println("Error on unit test: " + e);
	    e.printStackTrace();
	}

    }

}
